src/mqueryfront/src/QueryPage.js

```diff
@@ -90,10 +91,60 @@ class QueryPage extends Component {
         
     }
 
+
+    loadJob() {
+        const LIMIT = 50;
+
+        if (!this.state.qhash) {
+            return;
+        }
+
+        axios
+            .get(
+                API_URL +
+                    "/matches/" +
+                    this.state.qhash +
+                    "?offset=" +
+                    0 +
+                    "&limit=" +
+                    LIMIT
+            )
+            .then((response) => {
+                let newShouldRequest = true;
+
+                if (
+                    ["done", "cancelled", "failed", "expired"].indexOf(
+                        response.data.job.status
+                    ) !== -1
+                ) {
+                    if (response.data.job.files_processed >= response.data.job.total_files) {
+                        newShouldRequest = false;
+                    }
+                }
+
+                this.setState({
+                    job: response.data.job,
+                });
+
+                if (newShouldRequest) {
+                    let nextTimeout =
+                        response.data.matches.length >= LIMIT ? 50 : 1000;
+                    this.timeout = setTimeout(
+                        () => this.loadJob(),
+                        nextTimeout
+                    );
+                }
```

Comment:
All my suggestions in one comment:

```suggestion
                let job = response.data.job;
                this.setState({
                    job: job,
                    matches: response.data.matches,
                });

                let doneStatuses = ["done", "cancelled", "failed", "expired"];
                let isDone = doneStatuses.indexOf(job.status) !== -1;
                let processedAll = job.files_processed >= job.total_files;
                if (isDone && processedAll) {
                    return;
                }

                this.timeout = setTimeout(() => this.loadJob(), 1000);
```

Reply:
After this rewrite it also shows a bug (that I think was there even before your changes) - if the job is cancelled/failed but it didn't process all files, mquery will keep querying the backend. Maybe it should be `||` instead of `&&`?